command/debug/debug.go

```diff
@@ -515,20 +515,33 @@ func (c *cmd) captureLogs(timestampDir string) error {
 	}
 }
 
-func (c *cmd) captureMetrics(timestampDir string) error {
-
-	metrics, err := c.client.Agent().Metrics()
+func (c *cmd) captureMetrics(ctx context.Context, timestampDir string) error {
+	stream, err := c.client.Agent().MetricsStream(ctx)
 	if err != nil {
 		return err
 	}
+	defer stream.Close()
 
-	marshaled, err := json.MarshalIndent(metrics, "", "\t")
+	filename := fmt.Sprintf("%s/%s.json", timestampDir, "metrics")
+	fh, err := os.Create(filename)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to create metrics file: %w", err)
 	}
+	defer fh.Close()
 
-	err = ioutil.WriteFile(fmt.Sprintf("%s/%s.json", timestampDir, "metrics"), marshaled, 0644)
-	return err
+	decoder := json.NewDecoder(stream)
+	encoder := json.NewEncoder(fh)
+	// TODO: is More() correct here?
+	for decoder.More() {
+		var raw interface{}
+		if err := decoder.Decode(&raw); err != nil {
+			return fmt.Errorf("failed to decode metrics: %w", err)
+		}
+		if err := encoder.Encode(raw); err != nil {
```

Comment:
There is an open issue [7872](https://github.com/golang/go/issues/7872) about the json encoder buffering the whole file and not flushing until close time. That mean that we will not flush to disk before close, not sure if it's an issue in case of app crash debug

Reply:
In this case I think it is safe because the API returns multiple JSON objects. I believe the encoder only buffers until it reaches an object, so this will write each one separately. Each call to `Encode` will reach a full JSON object, and write it to the file, then the loop will continue to wait for the next full JSON object.

This can be replaced with an `io.Copy` though, if we  make the other change.